registration/app/registration/controllers/Main.scala

```diff
@@ -75,6 +77,46 @@ final class Main(
     result.map(processResponse)
   }
 
+  def registrationsByTopic(topic: Topic, cursors: Option[CursorSet]): Action[AnyContent] = Action.async {
+    val isFirstPage = cursors.isEmpty
+
+    val registrarResults = registrarProvider.withAllRegistrars { registrar =>
+      val cursorForProvider = cursors.flatMap(_.providerCursor(registrar.providerIdentifier))
+
+      if (isFirstPage || cursorForProvider.isDefined)
+        registrar.findRegistrations(topic, cursorForProvider.map(_.cursor)).map(_.toOption)
+      else
+        Future.successful(None)
+    }
+    Future.sequence(registrarResults).map { pageAttempts =>
+      val pages = pageAttempts.flatten
+
+      val results = pages.flatMap(_.results)
+
+      val cursor =
+        Some(pages.flatMap(_.cursor))
+          .filter(_.nonEmpty)
+          .map(CursorSet.apply)
+
+      Ok(Json.toJson(Paginated(results, cursor)))
+    }
+  }
+
+  def registrationsByDeviceToken(platform: Platform, deviceToken: String): Action[AnyContent] = Action.async {
+    val result = for {
+      registrar <- XorT.fromXor[Future](registrarProvider.registrarFor(platform))
+      registrations <- XorT(registrar.findRegistrations(deviceToken): Future[Xor[NotificationsError, List[StoredRegistration]]])
+    } yield registrations
+    result.fold(processErrors, res => Ok(Json.toJson(res)))
+  }
+
+  def registrationsByUdid(udid: UniqueDeviceIdentifier): Action[AnyContent] = Action.async {
+    val results = registrarProvider.withAllRegistrars { registrar =>
+      registrar.findRegistrations(udid)
+    }
+    Future.sequence(results).map(_.flatMap(_.toList.flatMap(_.results))).map(res => Ok(Json.toJson(res)))
```

Comment:
Maybe you are doing too much on a single line. What do you think about creating an function.